### DISCUSSION TITLE
Will this work with DirectX / COM?

### DISCUSSION THREAD

**Discussion by atcarter714** [microsoft/win32metadata#1294] [General] [answered]
The elephant in the room ... the most fun and interesting part of the Windows SDK that's been curiously overlooked by Project Reunion and that we've never had much official support for other than quickly abandoned bindings like MDX, great but short-lived products like XNA and aging community projects like SlimDX and SharpDX that are now cold to the touch. Game developers overwhelmingly want to be able to use DirectX _directly_ from .NET for 3D apps, games, engines, editors, etc, and creating an interop binding is complicated and a full-time job for a single person or small team. That's how I ended up here and found this ... it sounds like the capability exists but it's hard to find and research and I don't think many people know much about WinRT, .winmd files or the likes. I'm certainly just finding out about it myself.

So, is this going to work with DirectX's COM interfaces and functions? Is there already a C# projection for DirectX, at least for D3D12, the latest DXGI, etc?

**Comment by kennykerr** [comment, marked as answer]
The win32 metadata includes full coverage of DirectX including Direct3D 12. You can find examples of its application in Rust here:

https://github.com/microsoft/windows-rs/tree/master/crates/samples

I don't know how well this is supported by other languages, but the metadata is at least sufficient to support these APIs.

**Reply by atcarter714** [reply]
Oh that is exciting to hear indeed, and from Mr. Kerr himself, no less! 🙂

So I just made a .NET 6 Console app and added the NuGet package. I just want to do something simple like enumerate my adapters and outputs with DXGI 1.6 in C# as a first test. I added a NativeMethods.txt file and I'm afraid I'm a little lost now ... is there more documentation on this to take it from here? Once I get rolling I can figure out the rest and work out any marshaling or other complications that might crop up (I have the old SharpDx source as a guide anyway). Just not quite sure how the usage of this incredible-sounding pre-release package goes. I'm really excited about this project though and the possibilities it offers. I have a full-time job (game dev) but this is something I'd like to support if it can fill the void in my life left since SharpDx became a haggard old man and died on me ...

**Reply by atcarter714** [reply]
To be more specific, what syntax is necessary to add COM interfaces and Win32 functions in this NativeMethods.txt? I imagine something like IDXGIFactoryX interfaces require a bit of additional finesse?

**Reply by kennykerr** [reply]
I believe @AArnott is in charge of the C# projection and should be able to help.

**Reply by atcarter714** [reply]
> I believe @AArnott is in charge of the C# projection and should be able to help.

Yeah, I notice I have namespaces now like ABI.Windows.Graphics.DirectX and stuff in my project. Not sure how complete this is or how to go about using it yet but I'm going to figure it out. I'm super interested and support this effort 110%. I'd definitely like to chat with @AArnott and find out how I can help in my free time, even if all I can do is add little bits and pieces or create a public usage example or something.

**Reply by atcarter714** [reply]
I think I'm getting somewhere now, I added d3d12.* and dxgi.* in NativeMethods.txt and now I see PInvoke.CreateDXGIFactory2 and a bunch of expected function calls. Oddly, it wants a System.object for "ppFactory" parameter on that one and many others rather than a pointer or reference to something. Not sure how that marshalling is supposed to work or if that's really how it "should" be ... maybe that requires some hand-editing and custom marshaling. But I would think we'd need to give it a pointer to an IntPtr or something. 

I'm gonna keep playing with it and seeing what I can do, and I'd love to find out anything else I can and do anything I can to help once I've gotten my bearings.

**Reply by kennykerr** [reply]
There's enough metadata to turn such functions into something nice and generic. For example, the metadata for `CreateDXGIFactory2` is as follows:

```C#
public unsafe static extern HRESULT CreateDXGIFactory2([In] uint Flags, [In][Const] Guid* riid, [Out][ComOutPtr] void** ppFactory);
```

And in Rust that turns into (without any hand-editing):

```rust
fn CreateDXGIFactory2<T: Interface>(flags: u32) -> Result<T>;
```

I imagine C# can perform the same kind of generic return type transformation, but @AArnott would have to confirm.

**Reply by atcarter714** [reply]
That would be the ideal thing, to hand-craft some of this stuff into nice, pretty C# and hide unsafe stuff. You could do something like this

```
public static T CreateDXGIFactory<T>( uint Flags, out HRESULT hr ) where T: IDXGIFactory { ... }
```

Or something like that ... maybe a couple of overloads in line with C# conventions. May require hiding an uglier P/Invoke version or some fancier marshalling, but yeah, that's exactly how I'd like to see it implemented! I'm sold on this, I think this is super cool and important and should be available to everyone in .NET. Look forward to learning more and seeing if I can do anything to help @AArnott further this effort.

EDIT: I think I'm just missing something in my NativeMethods.txt I should be specifying since I don't have any COM interfaces types in IntelliSense. 

EDIT AGAIN: OK, I'm really close now. Only part I can't figure out is how to get the IID_DXGIFactory7 GUID from within my C# code, other than manually copying that string out of dxgi1_6.h or something and using Guid.Parse (ugly, hard-coded nonsense!). There's no __uuidof(x) equivalent in C# that I'm aware of. Once I get this figured out, I can create a much prettier C# assembly with "safe" wrappers and neatly organized namespaces for everything, which is something I've wanted for a long, long time since SharpDX went the way of the dinosaurs. I'm really excited about this!

**Reply by AArnott** [reply]
@atcarter714 We should probably move this discussion over to https://github.com/microsoft/CsWin32/discussions

**Reply by atcarter714** [reply]
> @atcarter714 We should probably move this discussion over to https://github.com/microsoft/CsWin32/discussions

You're right, man. I can start a discussion there instead. I sort of got things working after a nap, so I'll see ya over there!
